File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.java
class Solution {
    String res = "";
    public String longestPalindrome(String s) {
        int n = s.length();
        
        for(int i=0; i<n; i++) {
            // Odd length
            checkPalindrom(s, i-1, i+1);

            // Even length
            checkPalindrom(s, i, i+1);
        }

        return res;
    }

    void checkPalindrom(String s, int l, int r) {
        int n = s.length();

        while(l >= 0 && r < n && s.charAt(l) == s.charAt(r)) {
            l--;
            r++;
        }

        int length = (r - l - 1);
        if(length > res.length()) {
            res = s.substring(l+1, r);
        }
    }
}

File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        length = len(s)
        ans = ''
        dp = [[False for _ in range(length)] for _ in range(length)]
        
        for i in range(length - 1, -1, -1):
            for j in range(i, length):
                if i == j:
                    dp[i][j] = True
                elif j == i + 1:
                    dp[i][j] = s[i] == s[j]
                elif s[i] == s[j] and dp[i+1][j-1]:
                    dp[i][j] = True

                if dp[i][j] and (j - i + 1) > len(ans):
                    ans = s[i:j+1]

        return ans                    



        